### ISSUE TITLE
Linking error for an @EnableReflectiveInstantiation class inside a lambda in Scala 2.11 and 2.10

### ISSUE THREAD

**Issue by cheeseng** [scala-js/scala-js#3227]
Hi, 

Currently in ScalaTest 3.2.x branch: 

https://github.com/scalatest/scalatest/tree/3.2.x

If you do: 

```
> sbt scalatestAppJS/clean
> sbt scalatestTestJS/test:compile
> sbt scalatestTestJS/test
```

There is no problem to build, link and run the tests under scala 2.12 (the default scala version we are building).

But if we use: 

```
> sbt ++2.11.12 scalatestAppJS/clean
> sbt ++2.11.12 scalatestTestJS/test:compile
> sbt ++2.11.12 scalatestTestJS/test
```

It will blow up with a weird linking error that we cannot figure out what's wrong, and same behavior noticed in scala 2.10.7 also, any idea what's wrong with scala-js for scala 2.11 and 2.10?

Thanks.

cc @sjrd  @bvenners

**Comment by sjrd** [comment]
Could you post the linking error here, please?

**Comment by cheeseng** [comment]
@sjrd Sorry, here's the linking error: 

```
[info] Fast optimizing /home/cheeseng/git/scalatest/scalatest-test.js/target/scala-2.11/scalatesttestjs-test-fastopt.js
[error] Referring to non-existent class org.scalatest.NonImplicitAssertionsSuite$$anonfun$1
[error]   called from org.scalatest.NonImplicitAssertionsSuite$$anonfun$1$MySpec$1.<clinit>()
[error]   called from core module analyzer
[error] Referring to non-existent class org.scalatest.NonImplicitAssertionsSuite$$anonfun$1
[error]   called from org.scalatest.NonImplicitAssertionsSuite$$anonfun$1$MySpec$1.<clinit>()
[error]   called from core module analyzer
[error] Referring to non-existent class org.scalatest.NonImplicitAssertionsSuite$$anonfun$2
[error]   called from org.scalatest.NonImplicitAssertionsSuite$$anonfun$2$MySpec$2.<clinit>()
[error]   called from core module analyzer
[error] Referring to non-existent class org.scalatest.NonImplicitAssertionsSuite$$anonfun$2
[error]   called from org.scalatest.NonImplicitAssertionsSuite$$anonfun$2$MySpec$2.<clinit>()
[error]   called from core module analyzer
org.scalajs.core.tools.linker.LinkingException: There were linking errors
	at org.scalajs.core.tools.linker.frontend.BaseLinker.linkInternal(BaseLinker.scala:160)
	at org.scalajs.core.tools.linker.frontend.BaseLinker.linkInternal(BaseLinker.scala:107)
	at org.scalajs.core.tools.linker.frontend.LinkerFrontend$$anonfun$4.apply(LinkerFrontend.scala:62)
	at org.scalajs.core.tools.linker.frontend.LinkerFrontend$$anonfun$4.apply(LinkerFrontend.scala:62)
	at org.scalajs.core.tools.logging.Logger$class.time(Logger.scala:28)
	at org.scalajs.sbtplugin.Loggers$SbtLoggerWrapper.time(Loggers.scala:7)
	at org.scalajs.core.tools.linker.frontend.LinkerFrontend.link(LinkerFrontend.scala:61)
	at org.scalajs.core.tools.linker.Linker$$anonfun$link$1.apply$mcV$sp(Linker.scala:51)
	at org.scalajs.core.tools.linker.Linker$$anonfun$link$1.apply(Linker.scala:50)
	at org.scalajs.core.tools.linker.Linker$$anonfun$link$1.apply(Linker.scala:50)
	at org.scalajs.core.tools.linker.Linker.guard(Linker.scala:69)
	at org.scalajs.core.tools.linker.Linker.link(Linker.scala:50)
	at org.scalajs.core.tools.linker.ClearableLinker$$anonfun$link$1.apply(ClearableLinker.scala:52)
	at org.scalajs.core.tools.linker.ClearableLinker$$anonfun$link$1.apply(ClearableLinker.scala:52)
	at org.scalajs.core.tools.linker.ClearableLinker.linkerOp(ClearableLinker.scala:63)
	at org.scalajs.core.tools.linker.ClearableLinker.link(ClearableLinker.scala:52)
	at org.scalajs.sbtplugin.ScalaJSPluginInternal$$anonfun$org$scalajs$sbtplugin$ScalaJSPluginInternal$$scalaJSStageSettings$6$$anonfun$apply$8$$anonfun$apply$9.apply(ScalaJSPluginInternal.scala:308)
	at org.scalajs.sbtplugin.ScalaJSPluginInternal$$anonfun$org$scalajs$sbtplugin$ScalaJSPluginInternal$$scalaJSStageSettings$6$$anonfun$apply$8$$anonfun$apply$9.apply(ScalaJSPluginInternal.scala:297)
	at sbt.FileFunction$$anonfun$cached$1.apply(Tracked.scala:253)
	at sbt.FileFunction$$anonfun$cached$1.apply(Tracked.scala:253)
	at sbt.FileFunction$$anonfun$cached$2$$anonfun$apply$3$$anonfun$apply$4.apply(Tracked.scala:267)
	at sbt.FileFunction$$anonfun$cached$2$$anonfun$apply$3$$anonfun$apply$4.apply(Tracked.scala:263)
	at sbt.Difference.apply(Tracked.scala:224)
	at sbt.Difference.apply(Tracked.scala:206)
	at sbt.FileFunction$$anonfun$cached$2$$anonfun$apply$3.apply(Tracked.scala:263)
	at sbt.FileFunction$$anonfun$cached$2$$anonfun$apply$3.apply(Tracked.scala:262)
	at sbt.Difference.apply(Tracked.scala:224)
	at sbt.Difference.apply(Tracked.scala:200)
	at sbt.FileFunction$$anonfun$cached$2.apply(Tracked.scala:262)
	at sbt.FileFunction$$anonfun$cached$2.apply(Tracked.scala:260)
	at org.scalajs.sbtplugin.ScalaJSPluginInternal$$anonfun$org$scalajs$sbtplugin$ScalaJSPluginInternal$$scalaJSStageSettings$6$$anonfun$apply$8.apply(ScalaJSPluginInternal.scala:315)
	at org.scalajs.sbtplugin.ScalaJSPluginInternal$$anonfun$org$scalajs$sbtplugin$ScalaJSPluginInternal$$scalaJSStageSettings$6$$anonfun$apply$8.apply(ScalaJSPluginInternal.scala:291)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[error] (scalatestTestJS/test:fastOptJS) org.scalajs.core.tools.linker.LinkingException: There were linking errors
[error] Total time: 36 s, completed Dec 18, 2017 9:22:05 AM
```

NonImplicitAssertionsSuite is a test class that we are pretty sure is included, and the same error did not happen in Scala 2.12, which runs the test just fine.

**Comment by sjrd** [comment]
Minimization:
```scala
package helloworld

import scala.scalajs.js
import scala.scalajs.reflect.annotation._

object HelloWorld {
  def test(): Unit = {
    { () =>
      @EnableReflectiveInstantiation
      class Foo
    }
  }
}
```
The generated .sjsir files are:
```scala
> helloworld/scalajsp helloworld/HelloWorld$.sjsir
module class Lhelloworld_HelloWorld$ extends O {
  def test__V() {
    new sjsr_AnonFunction0().init___sjs_js_Function0((lambda<>() = {
      (void 0)
    }))
  }
  def init___() {
    this.O::init___();
    mod:Lhelloworld_HelloWorld$<-this
  }
}
```
```scala
> helloworld/scalajsp helloworld/HelloWorld$$anonfun$test$1$Foo$1.sjsir
class Lhelloworld_HelloWorld$$anonfun$test$1$Foo$1 extends O {
  val $$outer$1: Lhelloworld_HelloWorld$$anonfun$test$1
  def helloworld$HelloWorld$$anonfun$Foo$$$outer__Lhelloworld_HelloWorld$$anonfun$test$1(): Lhelloworld_HelloWorld$$anonfun$test$1 = {
    this.$$outer$1
  }
  def init___Lhelloworld_HelloWorld$$anonfun$test$1($$outer: Lhelloworld_HelloWorld$$anonfun$test$1) {
    if (($$outer === null)) {
      throw mod:sjsr_package$.unwrapJavaScriptException__jl_Throwable__O(null)
    } else {
      this.$$outer$1 = $$outer
    };
    this.O::init___()
  }
  static def clinit___() {
    mod:sjs_reflect_Reflect$.registerInstantiatableClass__T__jl_Class__sjs_js_Array__V("helloworld.HelloWorld$$anonfun$test$1$Foo$1", classOf[Lhelloworld_HelloWorld$$anonfun$test$1$Foo$1], [[[classOf[Lhelloworld_HelloWorld$$anonfun$test$1]], (lambda<>(arg$outer: any) = {
      new Lhelloworld_HelloWorld$$anonfun$test$1$Foo$1().init___Lhelloworld_HelloWorld$$anonfun$test$1(arg$outer.asInstanceOf[Lhelloworld_HelloWorld$$anonfun$test$1])
    })]])
  }
}
```
Note that there is no `helloworld/HelloWorld$$anonfun$test$1.sjsir`, although it is referred to in a `classOf` because it is an argument type of the local `class Foo`'s constructor, which is reflectively instantiatable.

Fun fact, if we add a (bare) `new Foo` in the lambda, the problem does not appear anymore, because the lambda cannot be optimized as a JS function anymore, due to how `lambdalift` altered things. We can even derive a workaround from that, by then wrapping that `new Foo` in some dead-code, like so:
```scala
  def test(): Unit = {
    { () =>
      @EnableReflectiveInstantiation
      class Foo

      { () => new Foo }
    }
  }
```
This issue reminds me of https://github.com/scala-js/scala-js/issues/3222, although here we have a Scala function class, for which we cannot simply emit an `AbstractJSType`. We need to generate a proper `Class`-kinded .sjsir file for the anonymous function class.

The issue does not appear in 2.12+ because lambdas are compiled very differently by scalac in that version.

Not sure yet how to fix it. I recommend using the workaround I showed above for now.

**Comment by sjrd** [comment]
@gzm0 Do you think this is worth fixing, given that:

* it only happens in 2.10 and 2.11,
* for a class with enabled reflective instantiation,
* that is declared inside a lambda, *and*
* that is not otherwise explicitly instantiated

?

The "use case" in ScalaTest was a bit contrived, as it was only testing that the class itself *compiles*, and it was put in a lambda only because it was in a `test(...) { ... }` block which takes by-name argument. It can be tested equally well outside the `test` block since it is only asserting that the piece of code compiles.

**Comment by gzm0** [comment]
Didn't we disallow to instantiate nested classes recently? TBH, I think it would be completely reasonable to disallow `@EnableReflectiveInstantiation` on local things entirely (nested can be argued about, since the outer injection semantics can be somewhat clear, for local things, there is most of the time not even a name for the type of the thing that would have to be injected into the constructor).

**Comment by sjrd** [comment]
I think you're referring to https://github.com/scala-js/scala-js/commit/e28ff5e0444c1410c2b44d4f819ea6b24966c570, in which we disabled reflective instantiation of nested *modules*. It might be reasonable to also disable it for *local* classes, indeed. Disabling it for inner classes would prevent legitimate and well-specified use cases, so I wouldn't do that.

**Comment by sjrd** [comment]
Yes, absolutely. Not to mention the *order* of captures if there are several. Local classes can definitely not be reliably instantiated reflectively.

**Comment by sjrd** [comment]
Fixed in c1083f3a011ba2bda62c9ab7860a849acd0494a3.
